core/imt/inc/ROOT/TExecutor.hxx

```diff
@@ -0,0 +1,494 @@
+///
+/// ###ROOT::Internal::TExecutor::Map
+/// The two possible usages of the Map method are:\n
+/// * `Map(F func, unsigned nTimes)`: func is executed nTimes with no arguments
+/// * `Map(F func, T& args)`: func is executed on each element of the collection of arguments args
+///
+/// For either signature, func is executed as many times as needed by a pool of
+/// nThreads threads; It defaults to the number of cores.\n
+/// A collection containing the result of each execution is returned.\n
+/// **Note:** the user is responsible for the deletion of any object that might
+/// be created upon execution of func, returned objects included. ROOT::::Internal::TExecutor never
+/// deletes what it returns, it simply forgets it.\n
+///
+/// \param func
+/// \parblock
+/// a lambda expression, an std::function, a loaded macro, a
+/// functor class or a function that takes zero arguments (for the first signature)
+/// or one (for the second signature).
+/// \endparblock
+/// \param args
+/// \parblock
+/// a standard vector, a ROOT::TSeq of integer type or an initializer list for the second signature.
+/// An integer only for the first.\n
+/// \endparblock
+///
+/// **Note:** in cases where the function to be executed takes more than
+/// zero/one argument but all are fixed except zero/one, the function can be wrapped
+/// in a lambda or via std::bind to give it the right signature.\n
+///
+/// #### Return value:
+/// An std::vector. The elements in the container
+/// will be the objects returned by func.
+///
+/// ### ROOT::Internal::TExecutor::MapReduce
+/// This set of methods behaves exactly like Map, but takes an additional
+/// function as a third argument. This function is applied to the set of
+/// objects returned by the corresponding Map execution to "squash" them
+/// to a single object. This function should be independent of the size of
+/// the vector returned by Map due to optimization of the number of chunks.
+///
+/// #### Examples:
+/// ~~~{.cpp}
+/// root[] ROOT::Internal::TExecutor pool; auto ten = pool.MapReduce([]() { return 1; }, 10, [](std::vector<int> v) { return std::accumulate(v.begin(), v.end(), 0); })
+/// root[] ROOT::Internal::TExecutor pool(ROOT::Internal::ExecutionPolicy::kMultiprocess); auto hist = pool.MapReduce(CreateAndFillHists, 10, PoolUtils::ReduceObjects);
+/// ~~~
+///
+//////////////////////////////////////////////////////////////////////////
+
+
+namespace ROOT{
+
+namespace Internal{
+class TExecutor: public TExecutorCRTP<TExecutor> {
+public:
+
+   /// \brief Class constructor. Sets the default execution policy and initializes the corresponding executor.
+   /// Defaults to multithreaded execution policy if ROOT is compiled with IMT=ON and IsImplicitMTEnabled. Otherwise it defaults to a serial execution policy
+   /// \param nProcessingUnits [optional] Number of parallel processing units, only taken into account if the execution policy is kMultithread
+   explicit TExecutor(unsigned nProcessingUnits = 0) :
```

Comment:
"Units" as in "chunks"? That's the "processing units" that come to my mind. What's the usual name for an abstraction of degree of concurrency / parallelism?

Reply:
Processing units as in (parallel) workers. I found parallel processing units a good name, but without the parallel maybe workers is better?